mesonbuild/modules/gobuild.py

```diff
@@ -0,0 +1,250 @@
+            "link_with",
+            ContainerTypeInfo(list, (build.SharedLibrary, build.StaticLibrary)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "tags",
+            ContainerTypeInfo(list, (str)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "native",
+            bool,
+            default=False,
+        ),
+        KwargInfo(
+            "gomod_tidy",
+            bool,
+            default=True,
+        ),
+    )
+    def project(
+        self,
+        state: ModuleState,
+        args: T.Tuple[
+            str, T.List[T.Union[mesonlib.FileOrString, build.GeneratedTypes]]
+        ],
+        kwargs: ProjectKwargs,
+    ) -> ModuleReturnValue:
+        is_cross_build = state.environment.is_cross_build(
+            mesonlib.MachineChoice.BUILD
+            if kwargs["native"]
+            else mesonlib.MachineChoice.HOST
+        )
+
+        cross_compilers = self.detect_tools(state, is_cross_build)
+
+        proj_name, arg_sources = args
+        all_sources = self.interpreter.source_strings_to_files(
+            list(itertools.chain(arg_sources, kwargs["sources"]))
+        )
+        go_sum_target = build.CustomTarget(
+            f"{proj_name}: go mod tidy",
+            state.subdir,
+            state.subproject,
+            state.environment,
+            [
+                self.go_exe,
+                "-C",
+                "@CURRENT_SOURCE_DIR@",
+                "mod",
+                "tidy" if kwargs["gomod_tidy"] else "verify",
+            ],
+            self.interpreter.source_strings_to_files(["go.mod"]),
+            [f"go-mod-tidy-{proj_name}.log"],
+            capture=True,
+            build_always_stale=True,
+        )
+
+        build_dir_abs = Path(state.environment.get_build_dir())
+        target_dir_abs = build_dir_abs / state.backend.get_target_dir(go_sum_target)
+
+        link_dirs = {
+            "-L" + (build_dir_abs / state.backend.get_target_dir(lib)).as_posix()
+            for lib in kwargs["link_with"]
+        }
+        link_targets = {"-l" + lib.name for lib in kwargs["link_with"]}
+
+        if not is_cross_build:
+            # TODO: always override environment variables CC and CXX for native builds.
+            extra_env = {}
+        else:
+            extra_env = asdict(cross_compilers)
+
+        exe_target = build.CustomTarget(
+            f"{proj_name}: go build",
+            state.subdir,
+            state.subproject,
+            state.environment,
+            [
+                self.go_exe,
+                "-C",
+                "@CURRENT_SOURCE_DIR@",
+                "build",
+                "-tags",
+                ",".join(kwargs["tags"]),
+                "-o",
+                f"{target_dir_abs.as_posix()}/@OUTPUT0@",
+            ],
+            all_sources + [go_sum_target],
+            [proj_name],
+            extra_depends=kwargs["link_with"],
```

Comment:
> Before using -buildmode=c-archive I had tried to use a regular go build but found that it could sometimes start building before the C dependencies were fully built (and therefore failed to link to them).

@odrling This draft PR provides the `gobuild.project(link_with: ...)` argument that ensures the linking stage is performed after all C/C++ libraries are built.